### ISSUE TITLE
No `used-before-assignment` for use of nested function before its definition

### ISSUE THREAD

**Issue by da-dada** [PyCQA/pylint#6812]
### Bug description
```python
def outer():
    inner()
    def inner():
        pass

outer()
```
Should raise `used-before-assignment` on use of `inner()` but doesn't.

### Command used

```shell
@echo off
rem see https://wiki.geany.org/howtos/check_python_code
set PYTHONPATH=%PYTHONPATH%;C:\Users\Me\Documents\Programming\Python\Mylib\My
rem set PYTHONPATH
set configfile=C:\Users\Me\Documents\Programming\Geany\BuildCommands\Python3\pylintrc
rem C:\Users\Me\AppData\Local\Programs\Python\Python38\Scripts\pylint --version
echo Using configfile %configfile%
echo   [without C0111(missing-docstring), C0103(invalid-name), W0201(attribute-defined-outside-init), W0703(broad-except), R1710(inconsistent-return-statements)
echo   generated-members=wx, additional-builtins=App, Envi; overgeneral-exceptions=None]
C:\Users\Me\AppData\Local\Programs\Python\Python39\Scripts\pylint --rcfile=%configfile% --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" %1
rem C:\Users\Me\AppData\Local\Programs\Python\Python39\Scripts\pylint --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" %1
```

### Pylint output

```shell
Using configfile C:\Users\Me\Documents\Programming\Geany\BuildCommands\Python3\pylintrc
  [without C0111(missing-docstring), C0103(invalid-name), W0201(attribute-defined-outside-init), W0703(broad-except), R1710(inconsistent-return-statements)
  generated-members=wx, additional-builtins=App, Envi; overgeneral-exceptions=None]
************* Module pylint_error
pylint_error.py:1: [C0104(disallowed-name), foo] Disallowed name "foo"
pylint_error.py:2: [C0321(multiple-statements), foo] More than one statement on a single line
Compilation finished successfully.
```

### Expected behavior

expected `used-before-assignment`

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.9.12 (tags/v3.9.12:b28265d, Mar 23 2022, 23:52:46) [MSC v.1929 64 bit (AMD64)]
```

### OS / Environment

Edition	Windows 10 Home
Version	21H1
Installed on	‎21.‎08.‎2020
OS build	19043.1706
Experience	Windows Feature Experience Pack 120.2212.4170.0

**Comment by da-dada** [comment]
whilst you are at it, using **nonlocal** like this _always_ dumps with **SyntaxError**

````python
def outer():
  nonlocal a
  a = 1
outer()
````

- pylint gives a success
- but if you comment out `a = 1` pylint also finds the missing binding
